### PR TITLE
[WIP] DataPipes Developer docs

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59053 [WIP] DataPipes Developer docs**
* #58597 [WIP] Example of DataPipes and DataFrames integration
* #58153 [WIP] Add notebook with DataPipes API examples
* #58152 [WIP] Add simple `fork` DataPipe
* #58151 [WIP] Decorate ZipDataPipe as `zip`
* #58150 [WIP] Add simple `groupby` DataPipe
* #58149 [WIP] Make `batch` DataPipe sensitive to batch_level argument
* #58148 [WIP] Add simple `unbatch` DataPipe
* #58147 [WIP] Make `filter` DataPipe sensitive to batch_level argument
* #58146 [WIP] Make `shuffle`  DataPipe sensitive to batch_level argument
* #58145 [DataLoader] Rename `transform` DataPipe to `legacy_transform`

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Notebook file https://github.com/pytorch/pytorch/blob/dd270009ff781d87c4711b94adc7c85df63877b6/torch/utils/data/datapipes_tutorial_dev_loaders.ipynb
